### ISSUE TITLE
Invoker Detection

### ISSUE THREAD

**Issue by fmontesi** [jolie/jolie#12]
It would be nice to be able to retrieve information about the invoker of an input port. This would be useful, e.g., for logging the IP address of an invoker.

In general, we have many types of communication channels, so there would be different types of information. For example, a socket could give an IP address, a local location would give the memory address of the invoker, and so on.

It looks to me, however, that this concept of "invoker information" makes sense for all channels. Hence we could augment all CommChannels with a new abstract method that we use to retrieve this information, and then each specific CommChannel implementation would have to provide its own implementation of the method (for sockets, local, unix sockets, bluetooth, etc.).

The question then becomes, how do we access this from the language? Any proposals?

**Comment by klag** [comment]
I would add a parameter in the protocol where it is possible to specify where to store the information. The same that happens for the user-agent in the http protocol.

Maybe we could define a set of parameters that come from the channel which are in common to all the protocols

**Comment by mawolf** [comment]
I experience the same need when implementing the RAFT algorithm. What about making it available in the "request" variable? LIke so:

[ some_action ( request ) (response) {
    println@Console(request.remote_address)() 
} ]

**Comment by klag** [comment]
The idea should be to preserve the code from the different used protocol.
If we insert in the request message some specific fields which depends on
the protocol
we will have a behaviour strictly related to the protocol and this could
rise problems during the deployment.

I think that an operation should declare all the information it is
requesting for working without any hidden fields.
Thus, the binding between the protocol-related information and the message
to forward to the operation
should be made in the protocol parameter scope.




2016-03-22 12:01 GMT+01:00 Martin Wolf <notifications@github.com>:

**Comment by fmontesi** [comment]
The behaviour will be dependent on the protocol configuration anyway, because it will access a variable that is set by the protocol. However, @mawolf's proposal also changes the interface, which I don't like. We should try to leave at least that independent from the protocol configuration. 

Using the solution we have in http however has its limitations. The parameter is set when you receive the message at the network level in the interpreter, not when the behaviour actually consumes it. So it works only when you do not have correlation sets and hence can receive messages inside of a process. 
Also, this is information we probably want to monitor, but the monitor only gets what's in the message. Maybe it's time to engineer message metadata both in the language and the interpreter? But, how would that work? 

I'm writing in a hurry, let me know if something is unclear. :-)

**Comment by klag** [comment]
We should deprecate code that depends on the parameters of the protocol.

If it is the case we must shift the problem on the architecture. As it
follows:

Case 1:
type OpRequest: void {
    .msg: string
}


 [ my_op()() {
         // processing of .msg
         // processing of invoker IP (it is an example)
}]

in this case the IP of the invoker is a parameter which comes with the
request message.
This means that the type of the request operation would be written as it
follows:

type OpRequest: void {
    .msg: string
    .invoker_IP: string
}

 [ my_op()() {
         // processing of .msg
         // processing of invoker IP (it is an example)
}]

In order to catch the IP from the protocol we must aggregate the operation
with a courier.

inputPort HttpPort {
Location: ....
Protocol: http { binding of the IP from the protocol to a variable }
Aggregates: MyPort
}

courier {
 // adding the IP before forwarding
}


The problem now is that the request message requires also the invoker_IP
which is declared in the type.
At the present this is not allowed by the courier and the type extension
primitive but if we add the possibility
to hide some fields from the aggregated type, such a result will be
achievable.


If we consider the courier as part of the deployment phase, it will be easy
to have a behaviour which is independent from the protocol.













2016-03-22 13:14 GMT+01:00 Fabrizio Montesi <notifications@github.com>:

**Comment by fmontesi** [comment]
Yes, that's also my belief, we should move the problem to the architecture.

Also, in general, I think it's far more interesting to make the language
powerful enough to handle this as a special case rather than having lots of
behaviour code that depends on specific protocols. In this way, we keep the
protocol-dependent code in the courier. It also gives more sense to having
couriers.

In general, we should have more operators for manipulating interfaces, not
just extension. These are basically type operators, so we could probably
make a thesis out of this by importing to Jolie the operators of type
algebras (extended to interfaces). I'll look into this.

Another interesting possibility is having courier for output ports. Any
motivating examples for this?

**Comment by fmontesi** [comment]
PS: we should also have a notion of type aliasing, like in other languages,
such that we can simply get surfaces directly in Jolie without having to
use external tools. One step at a time.. :-)

**Comment by mwallnoefer** [comment]
Good proposal but I would rather keep these meta-information accessible over the `MetaJolie` package. So I would opt against changing the request types since it could also be `void`. That said I share @klag opinion about not breaking existing interfaces regarding different port types and protocols.
